### ISSUE TITLE
rpc: remote clock monitoring fails to recognize excess offset

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#4884]
See:

http://52.91.194.28/cgi-bin/display.py?entry=20160305T161219#20160305T161219
http://52.91.194.28/cgi-bin/display.py?entry=20160305T161642#20160305T161642
http://52.91.194.28/cgi-bin/display.py?entry=20160305T162220#20160305T162220

The error basically says that what should be a linearizable history is not linearizable. This is happening while clock skews in excess of seconds are introduced randomly on each node. The nodes do not "commit suicide" as would be necessary to prevent these errors.

**Comment by tschottdorf** [comment]
just to clarify, this is a test which is running with `--linearizable`, correct?

**Comment by knz** [comment]
It doesn't matter. Error happens both with and without --linearizable.
The examples linked above don't use --linearizable, but the following does:

http://52.91.194.28/cgi-bin/display.py?entry=20160305T171032#20160305T171032

**Comment by tschottdorf** [comment]
I was asking because you called this a consistency error, but you're violating the MaxOffset you've set on the cluster, so this is really not a consistency error but a failure of the clock detection/suicide trigger (which can never work 100% reliable, but should work better than it does now). Just clarifying for other readers. I re-titled the issue.

**Comment by knz** [comment]
ok thanks!

**Comment by knz** [comment]
Tested again after #5107 merged in, and reducing `monitorInterval` to 1 second (instead of the default 30s). I confirm that the consistency error that motivated this issue happens much less often, and instead the nodes abort with an error more often. 

However with the default monitoring interval of 30s there is a 30 second interval where consistency errors can happen. I can see this happen experimentally. Two questions thus remain:

- would it be interesting to reduce the default clock offset monitoring interval?
- what statements do we want to make about consistency in the face of drifting clock? Is there a way to qualify which types of consistency errors may or may not happen?

**Comment by tamird** [comment]
I think the current interval-based monitoring is inherently flawed. We should really be checking the clocks in the request path and aborting eagerly.

**Comment by tamird** [comment]
OK, this is slightly better now that #5512 is in, but the solution is still not totally robust. We need to do something adaptive with maxOffset, but it's no longer a high priority. Unassigning myself since I'm not currently working on this.

@knz perhaps it's worth closing this issue since the reports here are (hopefully) no longer relevant.

**Comment by knz** [comment]
Thanks @tamird for the update.
I think however that the issue is still relevant even if the original test reports are not up to date. Namely as long as the issue is not fixed there is still a risk of db consistency errors in the presence of large skews. I agree however the issue is now somewhat lower priority than initially.

There are two ways forward really:
- either make skew detection rock solid (which may or may not be theoretically possible)
- document exuberantly that we disclaim any consistency guarantee if the users don't run NTP or equivalent clock synchronization with guaranteed  sub-maxoffset precision.

Should we move this issue to doc?

**Comment by tamird** [comment]
Yes, this should be a doc issue now. Improving the implementation is a
worthwhile pursuit, but it's going to be a complete rethinking and deserves
a separate issue.

**Comment by tamird** [comment]
@jseldess any movement on this one?

**Comment by jseldess** [comment]
@tamird, @knz, we do mention clock synchronization in our [recommended production settings](https://www.cockroachlabs.com/docs/recommended-production-settings.html#clock-synchronization), but the focus there is on transactions. Should we expand to somehow mention that large skews introduce a risk of db consistency errors? Can either of you help with a more precise message here?

**Comment by tamird** [comment]
Yeah that description is entirely too specific.

> Run NTP or other clock synchronization software on each machine. CockroachDB needs moderately accurate time; if the machines’ clocks drift too far apart, bad things will happen: in the worst case, consistency guarantees will be lost; in the best case, nodes will self-terminate to protect data consistency.

**Comment by jseldess** [comment]
Thanks, @tamird. Can you say more about what's actually happening in the best and worse cases?

**Comment by tamird** [comment]
Not sure what you mean.
